### ISSUE TITLE
Running <testsuite> in parallel

### ISSUE THREAD

**Issue by ThomasDelteil** [paratestphp/paratest#179]
# Running testsuite defined in the config in parallel

I wanted to open the discussion on the possibility to run testsuite in parallel using paratest.

By this I mean, giving each runner a specific testsuite to run, the testsuite would be defined in the phpunit.xml config. 

You might ask why. Well, I tried to run paratest on our test suite on a 16 core machine using 16 process.
It only brought down the run time from 4.55 min to 2.8 min compared to using only phpunit. Good but far from being 16 times better. I figured out that a lot of time was lost launching phpunit for every file. (even when using the --runner WrapperRunner). 
As a quick experiment, by creating 16 unoptimized arbitrary test suites and running them in background using phpunit. I managed to get all the test finished in 38s. (last suite taking 15+s than the previous one)

I wrote a python command line tool that takes the result from phpunit --log-json (time taken for each test), and a phpunit config file, and figure out an approximation of the optimal combination of test files to make up **N** `<testsuite>`, and then insert them in the configuration file alongside the original `<testsuite>`. The original testsuite has <exclude> tags with all the files of the N other `<testsuite>`. This is to ensure that we don't need to run the optimization step every time we add a new test. New tests being picked up by this original `<testsuite>`.

An example of the time taken for 16 suites generated from the json-output.

```
name: GeneratedSuite8, time:7.32033586502
name: GeneratedSuite6, time:7.32033967972
name: GeneratedSuite5, time:7.32033991814
name: GeneratedSuite14, time:7.32034683228
name: GeneratedSuite13, time:7.32034707069
name: GeneratedSuite10, time:7.32037878036
name: GeneratedSuite12, time:7.32040476799
name: GeneratedSuite15, time:7.32042336464
name: GeneratedSuite2, time:7.32046842575
name: GeneratedSuite4, time:7.3205037117
name: GeneratedSuite3, time:7.32056117058
name: GeneratedSuite11, time:7.32058119774
name: GeneratedSuite7, time:7.32060718536
name: GeneratedSuite9, time:7.32060742378
name: GeneratedSuite1, time:7.32061219215
name: GeneratedSuite0, time:11.5843603611 (made a up of a single very long test)
```

What do you think about this approach? Optimization + running suites in parallel?

**Comment by ThomasDelteil** [comment]
So early results:
after running the optimization to create 4 suites (my local machine has only 4 cores)

paratest **--parallel-suite true** --runner WrapperRunner -p 4 --configuration=build/phpunit_generated.xml
```
Time: 1 minute, Memory: 13.50Mb

OK (8700 tests, 46511 assertions)
```

running the same config file with **paratest**
paratest --runner WrapperRunner -p 4 --configuration=build/phpunit_generated.xml
```
Time: 1.48 minutes, Memory: 90.75Mb

OK (8700 tests, 46511 assertions)
```

Running with only **phpunit**
phpunit -c build/phpunit_generated.xml
```
Time: 2.32 minutes, Memory: 245.00Mb

OK (8700 tests, 46511 assertions)
```

**Comment by ThomasDelteil** [comment]
You can find the POC https://github.com/ThomasDelteil/paratest/tree/parallel-suite
and the python utility also a POC, I will probably port it to PHP as command for paratest 
https://github.com/ThomasDelteil/phpunit-ts-optimizer

**Comment by julianseeger** [comment]
Have you compared your results with a combination of `-f` and `--batch-max-size`?
Increasing the batch size to >> 1, multiple testmethods will be executed in one process, reducing the overhead of spawning processes. This will of course not result in a balanced minimal-process execution but should result in a much better result than using -f with the default batch size of 0.

**Comment by ThomasDelteil** [comment]
I tried, if I remember correctly it was quite much slower. Though i didn't try for -m >>> 1.  Correct me if I am wrong, - f does batch methods together but not from different test files or does it? This particular suite has 513 test files. So that makes it at least spawning 513 processes. I will benchmark it against paratest own tests and let you know the results.

**Comment by ThomasDelteil** [comment]
I tried to run my version on the paratest suite itself, here are the results (running on a MBP 4 cores)

### PHPUnit alone
`phpunit`
`Time: 29.42 seconds, Memory: 22.00Mb`

### Paratest functional mode
`bin/paratest -f -m 20 --phpunit=/usr/local/bin/phpunit`
`Time: 22.91 seconds, Memory: 7.75Mb`

### Paratest non-functional mode
`bin/paratest -c phpunit.xml.dist`
`Time: 23.73 seconds, Memory: 7.75Mb`

### Paratest non-functional mode WrapperRunner
`bin/paratest -c phpunit.xml.dist --runner WrapperRunner`
`Time: 21.81 seconds, Memory: 8.25Mb`

# Using optimisation + parallel-suite
`bin/paratest -c phpunit_generated_4.xml --parallel-suite true`
`Time: 19.91 seconds, Memory: 3.75Mb`

**Comment by celowsky** [comment]
@ThomasDelteil do you still have a working link for your POC referenced in your earlier comment? Both https://github.com/ThomasDelteil/phpunit-ts-optimizer and https://github.com/ThomasDelteil/paratest/tree/parallel-suite are giving 404's.

**Comment by tomasfejfar** [comment]
This is what I'm looking for. Different test suites are independent, but tests withing testsuite needs to be sequantial, because they use shared resource.

**Comment by tomasfejfar** [comment]
@celowsky https://github.com/ThomasDelteil/paratest/tree/parallel-suites

**Comment by tomasfejfar** [comment]
Couldn't make it work though and not sure how to debug this :(

```
> paratest --parallel-suite "-p" "2"

Running phpunit in 2 processes with W:\...\vendor\bin\phpunit

Configuration read from W:\...\phpunit.xml.dist



  [InvalidArgumentException]
  Log file C:\Users\tomasfejfar\AppData\Local\Temp\PT_46E5.tmp is empty. This means a PHPUnit process has crashed.


paratest [-p|--processes PROCESSES] [-f|--functional] [--no-test-tokens] [-h|--help] [--coverage-clover COVERAGE-CLOVER] [--coverage-html COVERAGE-HTML] [--coverage-php COVERAGE-PHP] [-m|--max-batch-size MAX-BATCH-SIZE] [--filter FILTER] [--parallel-suite] [--phpunit PHPUNIT] [--runner RUNNER] [--bootstrap BOOTSTRAP] [-c|--configuration CONFIGURATION] [-g|--group GROUP] [--exclude-group EXCLUDE-GROUP] [--stop-on-failure] [--log-junit LOG-JUNIT] [--colors] [--testsuite [TESTSUITE]] [--path PATH] [--] [<path>]

Script paratest --parallel-suite handling the paratest event returned with error code 1
```

**Comment by ThomasDelteil** [comment]
Sorry I haven't been working on that the last 3 years and I stopped working on `php` feel free to close. @tomasfejfar sorry I won't be able to assist you with that

**Comment by paslandau** [comment]
Can this be closed after  #349 is merged?
